### PR TITLE
debug: tracing: Add Segger RTT linker section options

### LINKED ISSUE #53544: Cannot see both bootloader and application RTT output
When attaching the SEGGER J-Link RTT Viewer to a nRF52 based target, in most cases only the logs from application are visible and not the logs from MCUboot, more seldom only logs from MCUboot.

Expected behaviour: both logs from MCUboot and from application are expected to be visible, like using serial or SWO.

Found solution:
The issue is due to the fact that the viewer or debugger looks in the target memory for RTT control block and buffer, searching for the "SEGGER RTT" signature. As MCUboot and application have them at different addresses, it either finds one or the other. The solution would be having them at same address for both and avoiding the application to re-initialize them if already initialized by MCUboot.
Confirmation and fixing hints can be found i.e. in https://devzone.nordicsemi.com/f/nordic-q-a/30310/easy-way-to-merge-bootloader-and-application-rtt-output
In the Nordic SDK this patch works well.

In Zephyr CONFIG_SEGGER_RTT_SECTION_DTCM defines a different section (`.dtcm_data`) for RTT control block and buffer, yet it seems that:
 * The linker file doesn't support this config and a `orphan section .dtcm_data` warning is given by the linker;
 * The `SEGGER_RTT.c` module needs patches to initialize properly the variables in `.dtcm_data' section and to not initialize them if already done by MCUboot.

I could verify that the two fixing above solve the issue and both logs are finally visible.
I didn't find any other open issue or PR mentioning it. I'll propose new PRs for the above changes, unless a better solution exists.

Impact: difficult to debug and diagnose proprietary DFU.

Environment:
 - OS: Windows 10 (Linux will be tested soon)
 - Toolchain: zephyr-sdk-0.14.2
 - commit 8e4137f14901b1fab360603070d341c3b58490d4

### PR DESCRIPTION
**Pull request by giansta**
Allows optionally placing Segger RTT data either in a specific
linker section that is located at RAM start, or in a specific linker
section defined by a memory region in DTS, as third and fourth
alternative to the DTCM section or the default data section.
This is useful to share the fixed address for different programs,
typically bootloader and application, and have seamless logging.

Signed-off-by: Giancarlo Stasi <giancarlo.stasi.co@gmail.com>

fixes #53544

### CODE REVIEW

**Comment by giansta** [comment]
Expected to fix https://github.com/zephyrproject-rtos/zephyr/issues/53544 with https://github.com/zephyrproject-rtos/segger/pull/17 in Segger repo

**Comment by microbuilder** [comment]
@nordic-krch Seems OK to me, but mind having a look since you maintain the Segger integration?

**Comment by giansta** [comment]
> If you mean that the approval depends on the approval of the segger PR, I totally agree :-)

**Comment by giansta** [comment]
> @nordic-krch Seems OK to me, but mind having a look since you maintain the Segger integration?

If you mean that the approval depends on the approval of the segger PR, I totally agree :-)

**Comment by giansta** [comment]
Changed after comments in https://github.com/zephyrproject-rtos/zephyr/issues/53544.
I believe that the first compliance error happened because I pushed the project segger after this push.
The 2nd and 3rd errors, on the other hand, look docker internal issues.

File: drivers/debug/Kconfig.rtt
Comment (nordic-krch) [inline drivers/debug/Kconfig.rtt:78]: don't we want `default y`? If bootloader is not used then it has no difference and when bootload is used then probably most users would like to get logs from bootloader and application.

**Review by nordic-krch — COMMENTED** [review]
Can you also update west so that it points to the changes in segger repo?

File: drivers/debug/Kconfig.rtt
Comment (giansta) [reply drivers/debug/Kconfig.rtt:79]: Agree, done.

**Comment by giansta** [comment]
> 



> Can you also update west so that it points to the changes in segger repo?

I already tried, but I got a compliance error with CI and I had to remove it. I'll try again.

**Comment by giansta** [comment]
> Can you also update west so that it points to the changes in segger repo?

It should be done now.

**Comment by nordicjm** [comment]
@nordic-krch can you re-review this please?

File: west.yml
Comment (carlescufi) [inline west.yml:212]: Please remove this, you need to send your change to the Zephyr repo instead

**Review by carlescufi — CHANGES_REQUESTED** [review]

File: west.yml
Comment (giansta) [reply west.yml]: Removed that commit. I expect I have to update `west.yml` only once that the segger repo PR  #17 is merged, right?

File: include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld
Comment (carlocaione) [inline include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:244]: What about using `zephyr_linker_sources_ifdef()` instead?

**Review by carlocaione — DISMISSED** [review]

File: include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld
Comment (giansta) [reply include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:244]: Isn't it a CMake function? Indeed I see it used only in CMakeLists.txt files, while linker file don't use it. Can it be used in linker files?

Comment (carlocaione) [reply include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:244]: No, it is used by CMake to embed linker script snippets.

Comment (giansta) [reply include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:244]: I've never done that way. On the other end, this linker script has similar sections with conditional compile switch, it was natural for me to do the same. Do you have an example?

**Comment by MarcianoPreciado** [comment]
We're so close on this, any chance these changes can get made?

File: include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld
Comment (giansta) [reply include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:244]: Would it be possible to separate this request in another PR?

File: include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld
Comment (MarekPieta) [reply include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:235]: That could be helpful also if a bootloader does not enable RTT at all (bootloader still may accidentally modify RAM area which application uses for the RTT). For the bootloader use-case, maybe we could allow to reserve memory without actually enabling the RTT support?

Comment (MarekPieta) [reply include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:235]: I also wonder if we could somehow share required size of the RTT related section among bootloader and application (configuration must be aligned to ensure interoperability, otherwise we may see some strange errors). Maybe we could somehow propagate the information e.g. using SysBuild already

File: drivers/console/rtt_console.c
Comment (sigvartmh) [inline drivers/console/rtt_console.c:44]: For another PR but could we make the channel configurable?

File: include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld
Comment (sigvartmh) [reply include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:244]: Here is how I did it when trying to implement the same as you have done here: https://github.com/sigvartmh/fw-nrfconnect-zephyr-1/commit/009b403829368e7a6687aa46a24d63f22d71a8b1#diff-7ab5276f0606a2bd1620a91984e170d371c28c4d7884774234ebb7340936ec06

Comment (sigvartmh) [reply include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:244]: Sorry I forgot to include the segger_rtt.ld in my dump of this :/ it is pretty much the same as the section you have there.

File: include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld
Comment (sigvartmh) [reply include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:235]: I'm not sure Zephyr's build system + linker have anyway of putting it at the end of RAM? 

One way of achieving this is to ensure that the RTT section is the last thing placed in RAM as then we could use some of the values of the KConfig to figure out the size of the partition then set the (Location Counter)[https://ftp.gnu.org/old-gnu/Manuals/ld-2.9.1/html_chapter/ld_3.html#SEC10] to be  `RTT_START_ADDR = RAM SIZE - SIZE of RTT Block.`

The reason why this must be placed last is that if we move the LC forward we cannot move it back in GNUs LD. 

> Side note: ARMs scatter file/ARMs linker does have this ability.

Comment (sigvartmh) [reply include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:235]: > That could be helpful also if a bootloader does not enable RTT at all (bootloader still may accidentally modify RAM area which application uses for the RTT). Alternatively, for the mentioned bootloader use-case, maybe we could allow to reserve memory without actually enabling the RTT support?

Agreed this would be useful ref. https://interrupt.memfault.com/blog/noinit-memory#retrofitting-a-noinit-region-when-you-cant-update-the-bootloader

Comment (nordicjm) [reply include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:235]: It would just be better to have a memory region setup in the dts and use that specifically for RTT. It should not just be assumed to be at the end of RAM, there are many reasons why it should not be placed there.

Comment (sigvartmh) [reply include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:235]: @nordicjm : Sure a separate linker memory section for RTT would work fine thanks for the input 👍 Do you know how this can be achieved to section off the RAM by any chance?  Is there a way to define a node that would cut off the end of ram. 

IMO it should be user selectable where you want to place your RTT section but at least having it BEFORE and AFTER Application/Bootloader RAM would be two sensible options.

@giansta : https://docs.zephyrproject.org/apidoc/latest/devicetree__regions_8h.html Whatever Docs I found on this.

File: drivers/debug/Kconfig.rtt
Comment (giansta) [reply drivers/debug/Kconfig.rtt]: I've added a `if CPU_CORTEX_M`, thanks.

File: include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld
Comment (giansta) [reply include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld]: Thanks for the suggestions and sorry for my delay. It should be done now.

File: include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld
Comment (giansta) [reply include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld]: I didn't want to add this at beginning, as the case for this PR (and for related issue #53544) was having a bootloader and app sharing the configuration.
But yes, the case with another bootloader where we want to keep the application logs between resets is also interesting. I've added a 2nd custom behaviour to use a DTS memory region that can be placed at end of RAM or at another address. I've tested it defining a region for my board. The two different choices are necessary, because of the snippet usage for the former one.
Anyway I wouldn't set it as default choice, because it needs a DTS definition, I kept the "normal" start of RAM option as default.

File: drivers/console/rtt_console.c
Comment (giansta) [reply drivers/console/rtt_console.c:44]: Even simple, I agree to do it in another PR :-)

**Comment by giansta** [comment]
Hello. Even if this is a low prio PR, I did some job on it time ago trying to address all requests.
Also, the [corresponding PR](https://github.com/zephyrproject-rtos/segger/pull/17) in segger repo is still open.
Is it possible to reopen and review it? I don't seem to have the rights to reopen it.

**Comment by nordicjm** [comment]
@giansta the segger PR needs rebasing

**Comment by giansta** [comment]
> @giansta the segger PR needs rebasing

Rebased, thanks!

**Review by sigvartmh — APPROVED** [review]

**Comment by nordicjm** [comment]
@ioannisg @bbolen @microbuilder @povergoing @nordic-krch @MarekPieta since this is a very old PR and gets merge conflicts can you review and approve if you find it acceptable so that the module update can be merged and this is able to be merged without encountering another merge conflict?

**Comment by kartben** [comment]
@nordic-krch would be good to try to get this and the associated Segger PR in for 3.7?

File: modules/segger/Kconfig
Comment (ithinuel) [inline modules/segger/Kconfig:92]: I'm not fully familiar with devicetree & its use in Zephyr.

Where is the link with this config and the DTS configuration. Would you have an example for it please?

Comment (nordicjm) [reply modules/segger/Kconfig:92]: Ping @giansta

Comment (giansta) [reply modules/segger/Kconfig:92]: This feature was requested [here](https://github.com/zephyrproject-rtos/zephyr/pull/53569#discussion_r1247899876).
The link is just the config `SEGGER_RTT_SECTION_CUSTOM_NAME` that is enabled both using `SEGGER_RTT_SECTION_CUSTOM` and `SEGGER_RTT_SECTION_CUSTOM_DTS_REGION`. It's used in `SEGGER_RTT_Conf.h` in the modules repo.

Defining the DTS is then system or board specific. For my nRF52 board I tested it defining in DTS a region like this:
```
/ {
	soc {
		sram_rtt: memory@20000800 {
			compatible = "zephyr,memory-region", "mmio-sram";
			zephyr,memory-region = "SRAM_RTT";
			reg = <0x2000F800 0x800>;
		};
	};
};
```
The section name is defined in the line `zephyr,memory-region = "SRAM_RTT";` and both application and MCUboot prj.conf had to define `CONFIG_SEGGER_RTT_SECTION_CUSTOM_NAME="SRAM_RTT"`.

**Review by cfriedt — APPROVED** [review]
Definite improvement

**Review by ithinuel — APPROVED** [review]

**Comment by cfriedt** [comment]
Currently building against `v3.6-branch`, I'm getting a weird warning 

```shell
warning: orphan section `.rtt_buff_data' from `modules/segger/libmodules__segger.a(SEGGER_RTT.c.obj)' being placed in section `.rtt_buff_data'
```

Was this addressed already? I thought I remember seeing something about it.

**Comment by cfriedt** [comment]
Was able to verify this working with a custom stm32 board, `tests/boot/test_mcuboot`

Also needed the wrapper [here](https://github.com/fab-uleuh/zephyr-twister-rtt), which would also be a good reason to build something like this into west.

There is a PR for that.

**Review by nordic-krch — APPROVED** [review]

**Review by nordicjm — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
@giansta thanks so much for your patience.

**Comment by giansta** [comment]
Thanks @carlescufi! Anyway, this PR requires even https://github.com/zephyrproject-rtos/segger/pull/17
I would expect it to be merged too.

**Comment by gchwier** [comment]
Hi, after merging this PR not working tests:
* https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/subsys/dfu/img_util

on nrf52840dk/nrf52840.
To reproduce:
`${ZEPHYR_BASE}/scripts/twister -p nrf52840dk/nrf52840 -T ${ZEPHYR_BASE}/tests/subsys/dfu/img_util -s dfu.image_util  --west-flash --device-testing --device-serial /dev/ttyACM0 -c -vv`
output:
```
DEBUG   - DEVICE: Running TESTSUITE img_util
DEBUG   - DEVICE: ===================================================================
DEBUG   - DEVICE: START - test_check_flash
DEBUG   - DEVICE: ASSERTION FAIL [0] @ WEST_TOPDIR/zephyr/lib/utils/onoff.c:283
DEBUG   - DEVICE: E: r0/a1:  0x00000004  r1/a2:  0x0000011b  r2/a3:  0x00000001
DEBUG   - DEVICE: E: r3/a4:  0x00000004 r12/ip:  0x00000040 r14/lr:  0x00000871
DEBUG   - DEVICE: E:  xpsr:  0x21000000
DEBUG   - DEVICE: E: Faulting instruction address (r15/pc): 0x000074d0
DEBUG   - DEVICE: E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
DEBUG   - DEVICE: E: Current thread: 0 (unknown)
DEBUG   - DEVICE: E: Halting system
DEBUG   - Timed out while monitoring 
```

bisect shows that commit: https://github.com/zephyrproject-rtos/zephyr/commit/711ed082a2c361311a06b2c76c604f7e03fa8ddf

**Comment by kartben** [comment]
@gchwier I think it might be because https://github.com/zephyrproject-rtos/segger/pull/17 still hasn't been merged? cc @carlescufi

**Comment by gchwier** [comment]
@kartben I've used that PR and looks that it does not help, still the same issue.

**Comment by brycewilkins** [comment]
@giansta I hope you don't mind me trying to help get your Segger repo PR merged! I've encountered the problems here also and would love to see this resolved. Great work!

**Pull request merged**
